testing/trino-tests/src/test/java/io/trino/security/TestAccessControl.java

```diff
@@ -387,12 +389,81 @@ public void testUpdateAccessControl()
     @Test
     public void testMergeAccessControl()
     {
-        assertAccessAllowed("""
-                MERGE INTO orders o USING region r ON (o.orderkey = r.regionkey)
-                WHEN MATCHED AND o.orderkey % 2 = 0 THEN DELETE
-                WHEN MATCHED AND o.orderkey % 2 = 1 THEN UPDATE SET orderkey = null
-                WHEN NOT MATCHED THEN INSERT VALUES (null, null, null, null, null, null, null, null, null)
-                """);
+        String catalogName = getSession().getCatalog().orElseThrow();
+        String schemaName = getSession().getSchema().orElseThrow();
+
+        String targetTable = "merge_nation_target_" + randomTableSuffix();
+        String targetName = format("%s.%s.%s", catalogName, schemaName, targetTable);
+        String sourceTable = "merge_nation_source_" + randomTableSuffix();
+        String sourceName = format("%s.%s.%s", catalogName, schemaName, sourceTable);
+
+        assertUpdate(format("CREATE TABLE %s (nation_name VARCHAR, region_name VARCHAR)", targetTable));
+
+        assertUpdate(format("INSERT INTO %s (nation_name, region_name) VALUES ('FRANCE', 'EUROPE'), ('ALGERIA', 'AFRICA'), ('GERMANY', 'EUROPE')", targetTable), 3);
+
+        assertUpdate(format("CREATE TABLE %s (nation_name VARCHAR, region_name VARCHAR)", sourceTable));
+
+        assertUpdate(format("INSERT INTO %s VALUES ('ALGERIA', 'AFRICA'), ('FRANCE', 'EUROPE'), ('GERMANY', 'EUROPE'), ('RUSSIA', 'EUROPE')", sourceTable), 4);
+
+        String baseMergeSql = format("MERGE INTO %s t USING %s s", targetTable, sourceTable) +
+                "    ON (t.nation_name = s.nation_name)";
+        String deleteCase = "" +
+                "    WHEN MATCHED AND t.nation_name > (SELECT name FROM tpch.tiny.region WHERE name = t.region_name AND name LIKE ('A%'))" +
+                "        THEN DELETE";
+        String updateCase = "" +
+                "    WHEN MATCHED AND t.nation_name = 'GERMANY'" +
+                "        THEN UPDATE SET nation_name = concat(s.nation_name, '_foo')";
+        String insertCase = "" +
+                "    WHEN NOT MATCHED AND s.region_name = 'EUROPE'" +
+                "        THEN INSERT VALUES(s.nation_name, (SELECT 'EUROPE'))";
+
+        // Show that without SELECT on the source table, the MERGE fails regardless of which case is included
+        for (String mergeCase : ImmutableList.of(deleteCase, updateCase, insertCase)) {
+            withPrivilegeDenied(sourceTable, SELECT_COLUMN, () ->
```

Comment:
Use `assertAccessDenied`, like the other tests in this class.

Reply:
Changed.